rudder-lang/src/logger.rs

```diff
@@ -2,71 +2,74 @@
 // SPDX-FileCopyrightText: 2019-2020 Normation SAS
 
 use crate::error::*;
+use lazy_static::lazy_static;
 use serde::{Deserialize, Serialize};
-use std::cell::RefCell;
+use std::sync::RwLock;
 use std::{fmt, str::FromStr};
 
 // to keep LOGS usage as safe as possible, make sure it is only called from this file
 // any outside module call is just a push that happens through macros
 // and a clone to get a safe clone of the global to work on from outside the module
 
-thread_local! {
-    pub static LOGS: RefCell<Logs> = RefCell::new(Logs {
+lazy_static! {
+    pub static ref LOGS: RwLock<Logs> = RwLock::new(Logs {
         logs: Vec::new(),
         log_level: LogLevel::Warn,
     });
 }
 
 pub fn clone_logs() -> Logs {
-    LOGS.with(|logs| logs.borrow().clone())
+    LOGS.read().expect("bug: failed to read logs").clone()
 }
 
 // pub static mut LOGS.log_level: LogLevel = LogLevel::Warn;
 pub fn set_global_log_level(log_level: LogLevel) {
-    LOGS.with(|logs| logs.borrow_mut().update_log_level(log_level))
+    LOGS.write()
+        .expect("bug: failed to update logs")
+        .update_log_level(log_level)
 }
 
 macro_rules! error {
     ($ ( $ arg : tt ) *) => ({
         use crate::logger::LOGS;
-        if LOGS.with(|logs| logs.borrow().is_level_included(LogLevel::Error)) {
-            LOGS.with(|logs| logs.borrow_mut().push(LogLevel::Error, format!( $ ( $ arg ) * )));
+        if LOGS.read().expect("bug: failed to read logs").is_level_included(LogLevel::Error) {
+            LOGS.write().expect("bug: failed to update logs").push(LogLevel::Error, format!( $ ( $ arg ) * ));
```

Comment:
Are you sure this work ? 
The read result here still holds (it will be freed after the closing brace) so the write call will wait indefinitely.